### ISSUE TITLE
Embedding API: How to store private data in a class instance?

### ISSUE THREAD

**Issue by verpeteren** [munificent/wren#156]
# Introduction

I read that the embedding is not fleshed out yet. But I am anxious to try this.
While browsing through wren's source, I could not find a way to do this but maybe I am missing something.

# Naive solution

What I want to do is something like this:

## wren 'user' source code
    // untested wren (pseudo) code
    webserver = new Webserver( "localhost", 8888, 60 )
    webserver.addDocumentRoot( "/static/(?<path>.*)", "../var/www/static" )

Looking at a wren.h, the following comes to mind:

## c embedding source code
    // untested c (pseudo ) code
    static void Wn_Webserver_Constructor( WrenVm * runtime ) {
        ObjClass * thisObj  = wrenGetArgumentObject( runtime, 0 );
        const char * hostName = wrenGetArgumentString( runtime, 1 );
        uint16_t portNr = (uint16_t)  wrenGetArgumentDouble( runtime, 2 );
        int timeout = (int)  wrenGetArgumentDouble( runtime, 3 );
        struct webserver_t ws = Webserver_New( hostName, portNr, timeout );
        if ( ws != NULL ) {   
            wrenSetPrivate( thisObj, (void*) ws );
            wrenReturnBool( runtime, true );
            return;
        }
        wrenReturnBool( runtime, false );
    }

    static void Wn_Webserver_AddDocumentRoot( WrenVm * runtime ) {
        ObjClass * thisObj  = wrenGetArgumentObject( runtime, 0 );
        struct webserver_t * ws = ( struct webserver_t *) wrenGetPrivate( thisObj );
        if ( ws != NULL ) {
            const char * pattern = wrenGetArgumentString( runtime, 1 );
            const char * folder = wrenGetArgumentString( runtime, 2 );
            Webserver_AddDocumentRoot( ws, pattern, folder );
            wrenReturnBool( runtime, true );
            return;
        }
        wrenReturnBool( runtime, false );
    }
   
    static void Wn_Webserver_Destructor( WrenVm * runtime ) {
        ObjClass * thisObj  = wrenGetArgumentObject( runtime, 0 );
        struct webserver_t * ws = ( struct webserver_t *) wrenGetPrivate( thisObj );
        if ( ws != NULL ) {
            Webserver_Delete( ws );
            wrenReturnBool( runtime, true );
            return;
        }
        wrenReturnBool( runtime, false );
    }

    void Wren_Run_MyStuff( WrenVM * runtime ) {
        wrenDefineStaticMethod( runtime, "webserver", "new", 2, Wn_Webserver_Constructor );
        wrenDefineMethod( runtime, "webserver", "addDocumentRoot", 2, Wn_Webserver_AddDocumentRoot);
        wrenDefineMethod( runtime, "webserver", "delete", 0, Wn_Webserver_Destructor );
        //.. compile the wren 'user' source code and process it.
    }

# Discussion

Now assume that Webserver_New Webserver_AddDocumentRoot  and Webserver_Delete are in a seperate h and c file and later on linked to the executable.

## Changes needed
   
That would mean that by adding the following 3 functions to wren would get us a long way.
    ObjClass *     wrenGetArgumentObject    ( runtime, int index );
    void          wrenGetPrivate        ( ObjClass * object );
    void          wrenSetPrivate        ( ObjClass * object, void * data );

## Unfortunately
   
    * When the wrenGetPrivate function is called, should it stay on this instance, or visit it's parents?
    * Will the wrenDefineStaticMethod 'new' and/or 'delete' function names be 'reserved'; Will it be assured that these are only called by the wren runtime and not in a user script?
   
### Set private with constructor
   
    Instead of:
    void          wrenSetPrivate        ( ObjClass * object, void * data );
   
    it could be:
    void          wrenSetPrivate        ( ObjClass * object, void * data, WrenForeignMethodFn *  method );
    and upon GC, the method is called.

### Defineclass
   
    The Api could be extended with another 2 functions like:
    void wrenDefineClass(  WrenVM * runtime, const char * className, WrenForeignMethodFn constructor, int arityConstructor, WrenForeignMethodFn destructor );  //  assuming that the destructor will have arity = 0 as it is only called by the GC
    void wrenDefineDestructor(WrenVM* vm, const char* className, WrenForeignMethodFn method, int arityConstructor );  //  to allow also different constructors
   
# TL;DR

How should I use wren's API to store arbitrary data in a class's instance?
I have no clue, but might be able to produce something if I know what direction to take.

# Epiloge

This issue was first posted on the mailing list. As Bob pointed out, it is a missing feature. Right now, Wren has no concept of "user data". Probably something roughly similar to what Lua does: http://www.lua.org/pil/28.1.html would be fine

Peter Reijnders

**Comment by munificent** [comment]
My thoughts will probably be disorganized here because I haven't put too much mental effort into this yet, but I figured as should at least put down what I have.

I'm hoping to avoid exposing very many types in Wren's public API. So things like `ObjClass` and such that are used internally in the VM are not visible to embedders. All they see is a functional API that just takes and returns vanilla C values.

That's in conflict with some of what you propose, but I don't think it's the end of the world. If I understand you correctly, the idea is that each instance has a single private `void*` pointer that the embedder can get and set.

That works, but I think we'll want to be more flexible. It would be more powerful if the embedder can have userdata objects that have arbitrarily-sized blobs of memory. Here's what I'm thinking. We add this function to the API:

```c
wrenDefineInstantiator(WrenVM* vm, const char* className,
                       WrenInstantiateFn instantiator);
```

("Instantiator" is lame. We'll have to come up with a better name.) If you call this and pass in some class name, it takes that's class's normal allocation behavior and replaces it with some code that calls your C function. Your function's signature is:

```c
size_t WrenInstantiateFn(WrenVM* vm)
```

Your just is to just return how many bytes of data you want your object to have.

Note that you don't have access to the constructor parameters. Construction in Wren goes in two steps: instantiation then initialization. Normally the instantiation step just creates an object of the proper size. Then the initialization step is where the contructor bodies run.

This C function is replacing the instantiation step. The separation here is necessary because of how constructors can call superclass constructors. You have to have a fully usable `this` before *any* constructor bodies start running, which means you have to allocate first.

So, if you call `wrenDefineInstantiator()`, you replace the normal class instantion process with your own. This means the class probably can't have any regular instance fields. All of the memory the class is owned by you, the embedder now.

To do something with that data, you'll define foreign methods as usual. In those, you'll call a new function:

```c
void* wrenGetArgumentData(WrenVM* vm, int index);
```

Again, the name is terrible. But this will look up the argument at `index` and, assuming its an instance of a class that you previously called `wrenDefineInstantiator()` for, this will give you back a pointer to the number of bytes you requested.

That's well and good, but you never actually got to *initialize* this memory, right? Well, for that, I *think* you should just be able to define `new` itself as a foreign method. In other words, make your class's constructor a foreign method. I think that will just work. If not, you could always define the constructor in Wren and have it invoke some other foreign "initialize" method you define.

The remaining piece is how it interacts with the GC. We'll likely need another function where you can define a finalizer for a class. That will get called before memory for the object is freed. This will probably be needed eventually, but finalizers are a pain in the ass and place annoying constraints on the GC, so I'd like to put this one off for as long as possible.

What do you think?

**Comment by verpeteren** [comment]
### Single private void *

Yes, I am looking for a way to *eventually* start a single private void* pointer to store in each instance.
It does not need to change after the 'construction' any more, but it must be free-ed when the class is destructed by the function that the embedder assigns it to.



As you are strictly exposing only c-things, this is a good idea indeed.

### Further questions

[ ] *Will the 'new' static method be called automatically when the object is intsantiated?*

> webserver = new Webserver( "localhost", 8888, 60 )

  or should it be

> webserver = Webserver.new( "localhost", 8888, 60 )

[ ] *What will happen if the 'new' is called again?*

 or should in be a reserved method name, that may not be called by the usercode?

>  w1 = new Webserver( "localhost", 8888, 60 )
>  w1 = w1.new( "10.0.0.16", 80, 60)

[ ] *A finalizer is for the void pointer in the instance is needed.*

 Bbut this probably won't change at runtime. I think the embedder could know how to clean up while he is writing the WrenInstantiateFn. It could even be:

> typedef void (*WrenDestructFn)(WrenVM* vm, void * privateData );
> 
> wrenDefineInstantiator(WrenVM* vm, const char* className, WrenInstantiateFn instantiator, WrenDestructFn destructor);

That would eliminate the need for a 'delete'. 

### Naming

Naming is one of those dificult things indeed. What about WrenDefineClass

**Comment by munificent** [comment]
> [ ] Will the 'new' static method be called automatically when the object is intsantiated?

Yes. It's technically an instance method (this it has access to `this` in its body), but in either case, when you do `new Foo(1, 2, 3)`, Wren compiles it to something approximately like:

```dart
var foo = Foo.instantiate
foo.new(1, 2, 3)`
```

I say approximately because if you hand-write this code, it won't work. The actual names of the "instantiate" and "new" methods aren't normal identifiers (they're ` instantiate` with a leading space and `new`, a reserved word).

The `instantiate` method creates a new, empty uninitialized object. Then the `new` method runs the constructor body on it.

With `wrenDefineInstantiator()`, I'm proposing that you would essentially provide a foreign function definition of `instantiate`.

> [ ] What will happen if the 'new' is called again?

Since `new` is a reserved word, you can't actually manually invoke a constructor on an existing object. `foo.new` is a parse error.

> [ ] A finalizer is for the void pointer in the instance is needed.

For foreign objects whose private data has a pointer, yeah, you'll almost always need a finalizer too. However, there are lots of possible foreign objects that wouldn't. You could, for example, define your own matrix or vector type. You might want to define them in C for performance. Since `wrenDefineInstantiator()` would let you allocate an object with enough memory for the matrix or vector elements right inline with the object, there'd be no need for a pointer. Since they're just numbers, there's nothing to free when the object is GCed.

> What about WrenDefineClass

I don't think that conveys the right idea. It's not that you're creating a class, it's that you're creating (or modifying an existing one) such that its state is foreign.

**Comment by gsmaverick** [comment]
This is bug is very apparent when trying to create Wren wrappers for C libraries. I tried to hack around on a SQLite interface in Wren the other day but there's no way to connect the SQLite data in C with an instance of a Wren class.

**Comment by munificent** [comment]
Yeah, I totally agree. This is a high priority.

I need to get a unit testing story in place for the embedding API. I started hacking on it a bit but I haven't gotten it fully working. Once that's going, it will be easier to add stuff like this. Right now, I'm hesitant to add things that aren't tested, so this is sort of gated on that.

**Comment by verpeteren** [comment]
This seems a chicken-and-egg-situation.

What would be the best way to proceed?
* Shall I create a fresh branch for this and add a test (XP-style)
* Can I create a test situation based on the things that you allready created?

**Comment by munificent** [comment]
If you'd like to start working on this in parallel, that's fine. In the meantime, I'll try to make some more progress on the unit testing support I started hacking on locally. Once that's in place, you'll be able to come back around and use that to write tests for this.

How does that sound?

**Comment by verpeteren** [comment]
Sounds cool. I am really curious to your work.
In the mean time I started with [what I understood so far](https://github.com/munificent/wren/compare/master...verpeteren:embedding-testcase#diff-62efc6d586eba350546ea412c91f46ddR6).

**Comment by munificent** [comment]
I finally carved out some time to think about this and [wrote up a little proposal](https://github.com/munificent/wren/issues/290). I think it's fairly similar to what you have here, modulo the changes in constructors and the embedding API that have happened since you wrote this. What do you think?

**Comment by munificent** [comment]
OK, now that foreign classes and finalizers are working, I'm calling this complete. I think they should cover your use cases here but let me know if you need anything else.
